File: dash-api/leaf.py
from dataclasses import dataclass
import json
import os

from . import config

@dataclass
class LeafData:
    is_plugged_in: bool
    is_charging: bool
    cruising_range_ac_off_miles: float
    cruising_range_ac_on_miles: float
    icon_path: str


LEAF_ICON_NOT_PLUGGED_IN = "not_plugged_in.png"
LEAF_ICON_PLUGGED_IN = "plugged_in.png"
LEAF_ICON_CHARGING = "charging.png"

def get_leaf_summary():
    # Get the leaf summary content from leaf-summary.json
    leaf_summary_file = os.path.join(config.dashboard_input_dir, "leaf-summary.json")

    if not os.path.isfile(leaf_summary_file):
        print("ERROR: leaf-summary.json does not exist")
        return {"error": "leaf-summary.json does not exist"}

    with open(leaf_summary_file) as f:
        leaf_summary = json.load(f)

    return leaf_summary

def get_leaf_icon(plugged_in: bool, charging: bool):
    if plugged_in:
        if charging:
            return LEAF_ICON_CHARGING
        return LEAF_ICON_PLUGGED_IN
    return LEAF_ICON_NOT_PLUGGED_IN


def get_leaf_data():
    leaf_summary = get_leaf_summary()

    plugged_in = leaf_summary["is_connected"]
    charging = leaf_summary["charging_status"] != "NOT_CHARGING"
    leaf_icon = get_leaf_icon(plugged_in, charging)
    leaf_data = LeafData(
        cruising_range_ac_off_miles=leaf_summary["cruising_range_ac_off_miles"],
        cruising_range_ac_on_miles=leaf_summary["cruising_range_ac_on_miles"],
        is_plugged_in=plugged_in,
        is_charging=charging,
        icon_path=os.path.join(config.leaf_image_dir, leaf_icon),
    )

    return leaf_data

File: dash-api-go/data/leaf.go
package data

import (
	"fmt"
	"time"
)

type UpdateDate time.Time

const updatedAtFormat = "2006-01-02 15:04:05"

func (u *UpdateDate) UnmarshalJSON(b []byte) error {
	if len(b) < 2 || b[0] != '"' || b[len(b)-1] != '"' {
		return fmt.Errorf("invalid time format: %s", b)
	}
	b = b[1 : len(b)-1]
	t, err := time.Parse(updatedAtFormat, string(b))
	if err != nil {
		return err
	}
	*u = UpdateDate(t)
	return nil
}
func (u UpdateDate) MarshalJSON() ([]byte, error) {
	return []byte(`"` + time.Time(u).Format(updatedAtFormat) + `"`), nil
}

type LeafData struct {
	UpdateDate              UpdateDate `json:"update_date"`
	IsConnected             bool       `json:"is_connected"`
	ChargingStatus          string     `json:"charging_status"`
	CruisingRangeAcOffMiles float32    `json:"cruising_range_ac_off_miles"`
	CruisingRangeAcOnMiles  float32    `json:"cruising_range_ac_on_miles"`
	IconPath                string     `json:"icon_path"`
}

const (
	LEAF_ICON_NOT_PLUGGED_IN = "not_plugged_in.png"
	LEAF_ICON_PLUGGED_IN     = "plugged_in.png"
	LEAF_ICON_CHARGING       = "charging.png"
)

func (leafData *LeafData) GetIconPath() string {
	if leafData.ChargingStatus != "NOT_CHARGING" {
		return LEAF_ICON_CHARGING
	} else if leafData.IsConnected {
		return LEAF_ICON_PLUGGED_IN
	} else {
		return LEAF_ICON_NOT_PLUGGED_IN
	}
}

func GetLeafData() (*LeafData, error) {
	leafData, err := JsonReadSharedLock[LeafData]("leaf-summary.json")
	if err != nil {
		return nil, err
	}
	leafData.IconPath = leafData.GetIconPath()
	return leafData, nil
}
